cmd/promtool/archive.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"os"
+)
+
+const filePerm = 0644
+
+type archiverConfig struct {
+	archiveName string
```

Comment:
If it's just one arg, omit the config struct and pass the name as a constructor param directly.

Reply:
Generally we tried out all of the following in Prometheus:

- long list of constructor arguments
- option structs (like here)
- functional arguments (https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)

In the end, just simple and plain constructor arguments turned out best for most cases because they make it crystal clear what needs to be passed and explicitly let the compiler know if you need a new / different argument. If you add a new option to config struct, it's often not well-defined whether it's a required or optional option, and someone using the API will not even notice during compile time that they are missing an argument.

Still, we sometimes use option structs when the argument list gets too long, but even with long argument lists we often just use normal arguments now.

Reply:
I see! Thanks for explanation!

Reply:
In this case, I'll change [httpClientConfig](https://github.com/chyeh/prometheus/blob/PROM-1551/cmd/promtool/http.go#L28) too.